src/gp/src/GPMSAOptions.C

```diff
@@ -271,6 +297,293 @@ GPMSAOptions::~GPMSAOptions()
+                                      (*m_experimentScenarios[i])[p]);
+            maxScenario[p] = std::max(maxScenario[p],
+                                      (*m_experimentScenarios[i])[p]);
+          }
+
+      for (unsigned int p=0; p != dimScenario; ++p)
+        if (m_autoscaleMinMaxAll ||
+            m_autoscaleMinMaxScenario.count(p))
+          {
+            if ((m_scenarioScaleMin.size() > p) &&
+                ((m_scenarioScaleMin[p] != 0) ||
+                 (m_scenarioScaleRange[p] != 1)))
+              queso_error_msg("Cannot autoscale and manually scale the same scenario parameter");
+
+            this->set_scenario_parameter_scaling(p, minScenario[p],
+                                                    maxScenario[p]);
+          }
+    }
+
+  if (m_autoscaleMinMaxAll || !m_autoscaleMinMaxUncertain.empty())
+    {
+      V maxUncertain(*m_simulationParameters[0]);
+
+      V minUncertain(*m_simulationParameters[0]);
+
+      for (unsigned int i=1; i < m_simulationParameters.size(); ++i)
+        for (unsigned int p=0; p != dimParameter; ++p)
+          {
+            minUncertain[p] = std::min(minUncertain[p],
+                                       (*m_simulationParameters[i])[p]);
+            maxUncertain[p] = std::max(maxUncertain[p],
+                                       (*m_simulationParameters[i])[p]);
+          }
+
+      for (unsigned int p=0; p != dimParameter; ++p)
+        if (m_autoscaleMinMaxAll ||
+            m_autoscaleMinMaxUncertain.count(p))
+          {
+            if ((m_uncertainScaleMin.size() > p) &&
+                ((m_uncertainScaleMin[p] != 0) ||
+                 (m_uncertainScaleRange[p] != 1)))
+              queso_error_msg("Cannot autoscale and manually scale the same uncertain parameter");
+
+            this->set_uncertain_parameter_scaling(p, minUncertain[p],
+                                                     maxUncertain[p]);
+          }
+    }
+
+
+  if (m_autoscaleMeanVarAll || !m_autoscaleMeanVarScenario.empty())
+    {
+      V meanScenario(*m_simulationScenarios[0]);
+
+      for (unsigned int i=1; i < m_simulationScenarios.size(); ++i)
+        meanScenario += *m_simulationScenarios[i];
+
+      for (unsigned int i=0; i < m_experimentScenarios.size(); ++i)
+        meanScenario += *m_experimentScenarios[i];
+
+      meanScenario /= (m_simulationScenarios.size() +
+                       m_experimentScenarios.size());
+
+      V varScenario(m_simulationScenarios[0]->env(),
+                    m_simulationScenarios[0]->map());
+
+      for (unsigned int i=0; i < m_simulationScenarios.size(); ++i)
+        {
+          V varTerm = *m_simulationScenarios[i] - meanScenario;
+          varTerm *= varTerm;
+          varScenario += varTerm;
+        }
+
+      for (unsigned int i=0; i < m_experimentScenarios.size(); ++i)
+        {
+          V varTerm = *m_experimentScenarios[i] - meanScenario;
+          varTerm *= varTerm;
+          varScenario += varTerm;
+        }
```

Comment:
These routines for calculating the variance aren't numerically stable.  You can use [this](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm) instead.